design-system/src/theme-provider.tsx

```diff
@@ -54,34 +57,50 @@ const validateTheme = (themeName?: string): ThemeName => {
   return 'default';
 };
 
-const ThemeProvider = (props: ThemeProviderProps) => {
-  const root = useRef<HTMLElement>(document.querySelector(':root'));
-  const localScopeElement = useRef<HTMLDivElement>(null);
-  const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
-
-  const changeTheme = useCallback((newTheme: string) => {
-    setTheme(validateTheme(newTheme));
-  }, []);
-
-  useLayoutEffect(() => {
-    const vars = toVars(themes[theme]);
-    const targetElement =
-      props.scope === 'local' ? localScopeElement.current : root.current;
-    Object.entries(vars).forEach(([key, value]) => {
-      targetElement?.style.setProperty(key, value);
-    });
-  }, [theme, props.scope]);
-
-  const value = useMemo(() => {
-    return { theme, changeTheme };
-  }, [theme, changeTheme]);
-
-  return (
-    <ThemeContext.Provider value={value}>
-      <div ref={localScopeElement}>{props.children}</div>
-    </ThemeContext.Provider>
-  );
-};
+// used to cover SSR builds (for instance in Gatsby)
+const isBrowser = typeof window !== 'undefined';
+
+const ThemeProvider = forwardRef<HTMLDivElement, ThemeProviderProps>(
+  (props, ref) => {
+    const rootRef = useRef<HTMLElement>(
+      isBrowser ? document.querySelector<HTMLElement>(':root') : null
+    );
+    const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
+
+    const changeTheme = useCallback((newTheme: string) => {
+      setTheme(validateTheme(newTheme));
+    }, []);
+
+    const localRef = ref as unknown as RefObject<HTMLDivElement>;
+
+    useEffect(() => {
```

Comment:
Nit: any particular reason for not using `useLayoutEffect`?

Reply:
I got back to using `useLayoutEffect`